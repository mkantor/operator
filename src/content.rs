extern crate handlebars;
use crate::lib::*;
use handlebars::{Context, Handlebars, RenderContext, Renderable};
use serde::Serialize;
use std::io;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{} from content directory path '{}'.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
  .content_directory_path.to_string_lossy()
)]
pub struct RegisteredTemplateParseError {
    source: handlebars::TemplateError,
    content_directory_path: PathBuf,
}

#[derive(Error, Debug)]
#[error(
  "Failed to parse template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct UnregisteredTemplateParseError {
    source: handlebars::TemplateError,
}

#[derive(Error, Debug)]
pub enum ContentLoadingError {
    #[error(transparent)]
    TemplateParseError(#[from] RegisteredTemplateParseError),

    #[error(
    "Input/output error when loading{} from content directory path '{}'.",
    .name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or(String::from(" content")),
    .content_directory_path.to_string_lossy()
  )]
    IOError {
        source: io::Error,
        content_directory_path: PathBuf,
        name: Option<String>,
    },
}

#[derive(Error, Debug)]
#[error(
  "Rendering failed for template{}.",
  .source.template_name.as_ref().map(|known_name| format!(" '{}'", known_name)).unwrap_or_default(),
)]
pub struct TemplateRenderError {
    #[from]
    source: handlebars::RenderError,
}

pub struct ContentEngine<'a> {
    template_registry: Handlebars<'a>,
}

impl<'a> ContentEngine<'a> {
    pub fn from_content_directory(
        content_directory_path: &'a Path,
    ) -> Result<Self, ContentLoadingError> {
        let template_registry = {
            let mut template_registry = Handlebars::new();
            template_registry
                .register_templates_directory(".hbs", content_directory_path)
                .map_err(|template_render_error| match template_render_error {
                    handlebars::TemplateFileError::TemplateError(source) => {
                        ContentLoadingError::TemplateParseError(RegisteredTemplateParseError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                        })
                    }
                    handlebars::TemplateFileError::IOError(source, original_name) => {
                        // Handlebars-rust will use an empty string when the error does not
                        // correspond to a specific path.
                        let name = if original_name.is_empty() {
                            None
                        } else {
                            Some(original_name)
                        };
                        ContentLoadingError::IOError {
                            source,
                            content_directory_path: PathBuf::from(content_directory_path),
                            name,
                        }
                    }
                })?;
            template_registry
        };

        Ok(ContentEngine { template_registry })
    }

    pub fn new_content(
        &self,
        handlebars_source: &str,
    ) -> Result<ContentItem, UnregisteredTemplateParseError> {
        let template = handlebars::Template::compile2(handlebars_source, true)
            .map_err(|source| UnregisteredTemplateParseError { source })?;
        Ok(ContentItem::UnregisteredTemplate {
            template_registry: &self.template_registry,
            template,
        })
    }
}

#[derive(Serialize)]
struct GluonRenderData {
    version: GluonVersion,
}

#[derive(Serialize)]
struct RenderData {
    gluon: GluonRenderData,
}

pub enum ContentItem<'a> {
    /// A named template that exists in the registry. This variant must only be
    /// constructed with names that have already been validated against the
    /// registry.
    RegisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        name_in_registry: &'a str,
    },

    /// An anonymous template for on-the-fly rendering.
    UnregisteredTemplate {
        template_registry: &'a Handlebars<'a>,
        template: handlebars::Template,
    },
}

impl Render for ContentItem<'_> {
    type RenderArgs = GluonVersion;
    type Error = TemplateRenderError;

    fn render(&self, gluon_version: GluonVersion) -> Result<String, Self::Error> {
        let render_data = RenderData {
            gluon: GluonRenderData {
                version: gluon_version,
            },
        };

        let rendered_content = match &self {
            ContentItem::RegisteredTemplate {
                template_registry,
                name_in_registry,
            } => template_registry.render(name_in_registry, &render_data)?,
            ContentItem::UnregisteredTemplate {
                template_registry,
                template,
            } => {
                let context = Context::wraps(render_data)?;
                let mut render_context = RenderContext::new(None);
                template.renders(template_registry, &context, &mut render_context)?
            }
        };

        Ok(rendered_content)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;

    #[test]
    fn content_engine_can_be_created_from_valid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_VALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_ok(),
                "Content engine could not be created from {}",
                path
            );
        }
    }

    #[test]
    fn content_engine_cannot_be_created_from_invalid_content_directory() {
        for &path in &CONTENT_DIRECTORY_PATHS_WITH_INVALID_CONTENTS {
            assert!(
                ContentEngine::from_content_directory(Path::new(path)).is_err(),
                "Content engine was successfully created from {}, but this should have failed",
                path
            );
        }
    }

    #[test]
    fn new_templates_can_be_rendered() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &(template, expected_output) in &VALID_TEMPLATES {
            let new_content = engine
                .new_content(template)
                .expect("Template could not be parsed");
            let rendered = new_content
                .render(GluonVersion("0.0.0"))
                .expect(&format!("Template rendering failed for `{}`", template,));
            assert_eq!(
                rendered,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                rendered,
            );
        }
    }

    #[test]
    fn new_content_fails_for_invalid_templates() {
        let engine = ContentEngine::from_content_directory(
            arbitrary_content_directory_path_with_valid_content(),
        )
        .expect("Content engine could not be created");

        for &template in &INVALID_TEMPLATES {
            let result = engine.new_content(template);

            assert!(
                result.is_err(),
                "Content was successfully created for invalid template `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn new_templates_can_reference_partials_from_content_directory() {
        let content_directory_path = example_path("valid/partials");
        let engine = ContentEngine::from_content_directory(&content_directory_path)
            .expect("Content engine could not be created");

        let template = "this is partial: {{> ab}}";
        let expected_output = "this is partial: a\nb\n\n";

        let new_content = engine
            .new_content(template)
            .expect("Template could not be parsed");
        let rendered = new_content
            .render(GluonVersion("0.0.0"))
            .expect(&format!("Template rendering failed for `{}`", template,));
        assert_eq!(
            rendered,
            expected_output,
            "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
            template,
            expected_output,
            rendered,
        );
    }
}
