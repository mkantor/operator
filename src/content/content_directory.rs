use std::env;
use std::fs::File;
use std::os::unix::fs::PermissionsExt;
use std::path;
use std::path::{Path, PathBuf};
use thiserror::Error;
use walkdir::WalkDir;

#[derive(Error, Debug)]
pub enum ContentDirectoryFromRootError {
    #[error("Unable to use directory root '{}': {}", .root.display(), .message)]
    InvalidRootPath { root: PathBuf, message: String },

    #[error("Unable to use directory root '{}': {}", .root.display(), .source)]
    WalkDirError {
        root: PathBuf,
        source: walkdir::Error,
    },

    #[error(transparent)]
    DirectoryEntryError(#[from] ContentFileError),
}

#[derive(Error, Debug)]
#[error("Content file error: {}", .0)]
pub struct ContentFileError(String);

/// A filesystem directory containing content.
pub struct ContentDirectory {
    files: Vec<ContentFile>,
    #[cfg(test)]
    root: PathBuf,
}

impl ContentDirectory {
    pub fn from_root<P: AsRef<Path>>(
        absolute_root: &P,
    ) -> Result<Self, ContentDirectoryFromRootError> {
        let absolute_root_path = absolute_root.as_ref();
        if !absolute_root_path.is_absolute() {
            return Err(ContentDirectoryFromRootError::InvalidRootPath {
                message: String::from("Root path must be absolute."),
                root: PathBuf::from(absolute_root_path),
            });
        }

        let mut files = Vec::new();
        let walker = WalkDir::new(absolute_root_path)
            .follow_links(true)
            .min_depth(1);
        for dir_entry_result in walker {
            let dir_entry = dir_entry_result.map_err(|walkdir_error| {
                ContentDirectoryFromRootError::WalkDirError {
                    source: walkdir_error,
                    root: PathBuf::from(absolute_root_path),
                }
            })?;
            {
                let entry_path = dir_entry.path().to_path_buf();
                if dir_entry.file_type().is_file() {
                    let content_file =
                        ContentFile::from_root_and_path(absolute_root_path, entry_path)
                            .map_err(ContentDirectoryFromRootError::from)?;
                    files.push(content_file);
                }
            }
        }

        Ok(ContentDirectory {
            files,
            #[cfg(test)]
            root: absolute_root_path.into(),
        })
    }

    #[cfg(test)]
    pub fn root(&self) -> &Path {
        &self.root
    }
}

pub struct ContentFile {
    absolute_path: String,
    relative_path: String,
    is_hidden: bool,
    is_executable: bool,
    relative_path_without_extensions: String,
    extensions: Vec<String>,
    file: File,
}
impl ContentFile {
    pub const PATH_SEPARATOR: char = '/';

    fn from_root_and_path(
        content_directory_root: &Path,
        absolute_content_file_path: PathBuf,
    ) -> Result<Self, ContentFileError> {
        if path::MAIN_SEPARATOR != Self::PATH_SEPARATOR {
            return Err(ContentFileError(format!(
                "Platforms that use '{}' as a path separator are not supported",
                path::MAIN_SEPARATOR
            )));
        }

        let root = match content_directory_root.to_str() {
            Some(unicode_root) => unicode_root,
            None => {
                return Err(ContentFileError(format!(
                    "Non-unicode directory root (path is similar to '{}')",
                    content_directory_root.display(),
                )))
            }
        };

        let absolute_path = String::from(
            absolute_content_file_path
                .to_str()
                .ok_or_else(|| ContentFileError(String::from("Path was not unicode.")))?,
        );

        let relative_path = absolute_content_file_path
            .strip_prefix(root)
            .map_err(|strip_prefix_error| {
                ContentFileError(format!(
                    "Content file path '{}' did not start with expected prefix '{}': {}",
                    absolute_content_file_path.display(),
                    root,
                    strip_prefix_error
                ))
            })?
            .to_str()
            .map(String::from)
            .ok_or_else(|| ContentFileError(String::from("Path was not unicode.")))?;

        let file = File::open(&absolute_content_file_path).map_err(|io_error| {
            ContentFileError(format!(
                "Unable to open file '{}' in '{}' for reading: {}",
                relative_path, root, io_error
            ))
        })?;

        let basename = absolute_content_file_path
            .file_name()
            .ok_or_else(|| {
                ContentFileError(format!(
                    "Unable to get basename of '{}' in '{}'",
                    relative_path, root,
                ))
            })?
            .to_str()
            .ok_or_else(|| ContentFileError(String::from("File had a non-unicode basename.")))?;

        // Conventions around hidden files, whether a file is executable, etc
        // differ across platforms. It wouldn't be hard to implement this, but
        // Operator does not currently run its CI checks on non-unix platforms
        // so it would be too easy to introduce regressions.
        let (extensions, is_hidden, is_executable) = if !cfg!(unix) {
            return Err(ContentFileError(format!(
                "Operator does not currently support your operating system ({})",
                env::consts::OS,
            )));
        } else {
            // If the basename begins with `.` its first chunk isn't considered an "extension".
            let non_extension_components = if basename.starts_with('.') { 2 } else { 1 };
            let extensions = basename
                .split('.')
                .skip(non_extension_components)
                .map(String::from)
                .collect::<Vec<String>>();

            // The file is hidden if any of its relative path components starts
            // with a dot.
            let is_hidden = relative_path.starts_with('.')
                || relative_path.contains(&format!("{}.", Self::PATH_SEPARATOR));

            let permissions = file
                .metadata()
                .map_err(|io_error| {
                    ContentFileError(format!(
                        "Unable to query metadata for content file '{}': {}",
                        absolute_content_file_path.display(),
                        io_error
                    ))
                })?
                .permissions();
            let is_executable = permissions.mode() & 0o111 != 0;

            (extensions, is_hidden, is_executable)
        };

        let relative_path_without_extensions = String::from({
            let extensions_len = extensions.iter().fold(0, |len, extension| {
                // Extra 1 is to count . in the extensions.
                len + extension.len() + 1
            });
            &relative_path[0..(relative_path.len() - extensions_len)]
        });

        Ok(ContentFile {
            absolute_path,
            relative_path,
            relative_path_without_extensions,
            extensions,
            file,
            is_hidden,
            is_executable,
        })
    }

    pub fn absolute_path(&self) -> &str {
        &self.absolute_path
    }

    pub fn relative_path(&self) -> &str {
        &self.relative_path
    }

    pub fn is_hidden(&self) -> bool {
        self.is_hidden
    }

    pub fn is_executable(&self) -> bool {
        self.is_executable
    }

    pub fn relative_path_without_extensions(&self) -> &str {
        &self.relative_path_without_extensions
    }

    pub fn extensions(&self) -> &[String] {
        &self.extensions
    }

    pub fn file(&self) -> &File {
        &self.file
    }

    pub fn into_file(self) -> File {
        self.file
    }
}

impl IntoIterator for ContentDirectory {
    type Item = ContentFile;
    type IntoIter = std::vec::IntoIter<Self::Item>;
    fn into_iter(self) -> Self::IntoIter {
        self.files.into_iter()
    }
}

impl<'a> IntoIterator for &'a ContentDirectory {
    type Item = &'a ContentFile;
    type IntoIter = std::slice::Iter<'a, ContentFile>;
    fn into_iter(self) -> Self::IntoIter {
        self.files.iter()
    }
}

#[cfg(test)]
impl PartialEq for ContentDirectory {
    fn eq(&self, other: &Self) -> bool {
        self.root() == other.root()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;
    use std::fs;

    #[test]
    fn directory_can_be_created_from_valid_root() {
        let path = fs::canonicalize("./src").expect("Canonicalizing path failed");
        let result = ContentDirectory::from_root(&path);
        assert!(
            result.is_ok(),
            "Unable to use directory at '{}': {}",
            path.display(),
            result.err().unwrap().to_string()
        );
    }

    #[test]
    fn directory_root_must_exist() {
        let result = ContentDirectory::from_root(&example_path("this/does/not/actually/exist"));
        assert!(
            result.is_err(),
            "Directory was successfully created from non-existent path",
        );
    }

    #[test]
    fn directory_root_must_be_absolute_path() {
        let non_absolute_path = "./src";
        let result = ContentDirectory::from_root(&non_absolute_path);
        assert!(
            result.is_err(),
            "ContentDirectory was successfully created from non-absolute path '{}', but this should have failed",
            non_absolute_path,
        );
    }
}
