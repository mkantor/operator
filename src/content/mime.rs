use mime::Mime;
use serde::{Serialize, Serializer};
use std::fmt;
use std::str::FromStr;
use thiserror::Error;

pub type MediaRange = Mime;

/// See [IETF RFC 2046](https://tools.ietf.org/html/rfc2046).
#[derive(Clone, Eq, Hash, PartialEq)]
pub struct MediaType(MediaRange);
impl MediaType {
    pub const APPLICATION_OCTET_STREAM: Self = Self(mime::APPLICATION_OCTET_STREAM);

    pub fn from_media_range(media_range: MediaRange) -> Option<MediaType> {
        if media_range.type_() == "*" || media_range.subtype() == "*" {
            None
        } else {
            Some(MediaType(media_range))
        }
    }

    pub fn is_within_media_range(&self, media_range: &MediaRange) -> bool {
        if media_range == &::mime::STAR_STAR {
            true
        } else if media_range.subtype() == "*" {
            self.0.type_() == media_range.type_()
        } else {
            self == media_range
        }
    }

    pub fn into_media_range(self) -> MediaRange {
        self.0
    }
}

#[derive(Error, Debug)]
#[error("Could not parse media type: {}", .0)]
pub struct MediaTypeFromStrError(String);

impl FromStr for MediaType {
    type Err = MediaTypeFromStrError;
    fn from_str(input: &str) -> Result<Self, Self::Err> {
        let mime = Mime::from_str(input)
            .map_err(|error| MediaTypeFromStrError(format!("Malformed media type: {}", error)))?;
        MediaType::from_media_range(mime).ok_or_else(|| {
            MediaTypeFromStrError(String::from(
                "Input is a valid media range, but not a specific media type",
            ))
        })
    }
}

impl fmt::Debug for MediaType {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        fmt::Debug::fmt(&self.0, formatter)
    }
}

impl fmt::Display for MediaType {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        fmt::Display::fmt(&self.0, formatter)
    }
}

impl Serialize for MediaType {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

impl PartialEq<MediaRange> for MediaType {
    fn eq(&self, other: &MediaRange) -> bool {
        &self.0 == other
    }
}

impl PartialEq<MediaType> for MediaRange {
    fn eq(&self, other: &MediaType) -> bool {
        self == &other.0
    }
}
