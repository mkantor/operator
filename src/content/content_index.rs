use super::content_engine::CanonicalRoute;
use crate::content_directory::ContentFile;
use serde::Serialize;
use std::collections::HashMap;
use thiserror::Error;

#[derive(Error, Debug)]
#[error("Failed to add route '{}' to index: {}", .failed_route, .message)]
pub struct ContentIndexUpdateError {
    failed_route: String,
    message: String,
}

/// A hierarchial tree mapping out content in the registry. Does not actually
/// contain content items, just their routes.
///
/// For example, given the following content directory:
///
/// ```text
/// content/
///   foo.txt
///   bar.html
///   bar/
///     plugh.md.hbs
///     baz/
///       quux.gif
/// ```
///
/// The content index would be:
///
/// ```yaml
/// foo: foo
/// bar: bar
/// bar/:
///   plugh: bar/plugh
///   baz/:
///     quux: bar/baz/quux
/// ```
#[derive(Clone, Serialize)]
#[serde(untagged)]
pub enum ContentIndex {
    File(CanonicalRoute),
    Directory(ContentIndexEntries),
}

#[derive(Clone, Serialize)]
pub struct ContentIndexEntries(HashMap<String, ContentIndex>);
impl ContentIndexEntries {
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    pub fn try_add<C: AsRef<str>>(
        &mut self,
        canonical_route: C,
    ) -> Result<(), ContentIndexUpdateError> {
        let (dirname_components, basename) = {
            let mut path_components = canonical_route.as_ref().split(ContentFile::PATH_SEPARATOR);
            let basename = path_components.next_back();
            (path_components, basename)
        };

        match basename {
            None => Ok(()), // Successfully inserted nothing! ðŸŽ‰
            Some(basename) => {
                // Navigate to the correct place in the index by iterating path
                // components (except the last one), creating new directories
                // along the way as needed (think mkdir -p).
                let mut node = self;
                for path_component in dirname_components {
                    let next_node = node
                        .0
                        // Non-leaf nodes in the index end with `/` (they
                        // represent directories).
                        .entry(format!("{}/", path_component))
                        .or_insert_with(|| ContentIndex::Directory(Self::new()));

                    node = match next_node {
                        ContentIndex::Directory(branch) => branch,
                        ContentIndex::File(CanonicalRoute(conficting_route)) => {
                            // Each component in dirname_components represents
                            // a directory along the path
                            return Err(ContentIndexUpdateError {
                            failed_route: String::from(canonical_route.as_ref()),
                              message: format!(
                                "There is already a file at '{}', but that needs to be a directory to accommodate the new route.",
                                conficting_route,
                              )
                            });
                        }
                    };
                }

                // Use the last path component to insert a file in the index.
                match node.0.get(basename) {
                    Some(existing_entry) => {
                        let entry_description = match existing_entry {
                            ContentIndex::Directory(..) => "directory",
                            ContentIndex::File(..) => "file",
                        };
                        Err(ContentIndexUpdateError {
                            failed_route: String::from(canonical_route.as_ref()),
                            message: format!(
                                "There is already a {} at '{}'.",
                                entry_description,
                                canonical_route.as_ref(),
                            ),
                        })
                    }
                    None => {
                        node.0.entry(String::from(basename)).or_insert_with(|| {
                            ContentIndex::File(CanonicalRoute::new(canonical_route))
                        });
                        Ok(())
                    }
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn index_has_the_correct_structure() {
        let mut index = ContentIndexEntries::new();
        index.try_add("foo").unwrap();
        index.try_add("bar").unwrap();
        index.try_add("bar/plugh").unwrap();
        index.try_add("bar/baz/quux").unwrap();

        let actual_json = serde_json::to_value(index).unwrap();
        let expected_json = json!({
            "foo": "foo",
            "bar": "bar",
            "bar/": {
              "plugh": "bar/plugh",
              "baz/": {
                "quux": "bar/baz/quux"
              }
            }
        });
        assert_eq!(actual_json, expected_json);
    }
}
