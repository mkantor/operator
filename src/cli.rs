use crate::content::*;
use crate::content_directory::ContentDirectory;
use crate::lib::*;
use mime::Mime;
use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum RenderCommandError {
    #[error("Failed to read input.")]
    ReadError { source: io::Error },

    #[error("Unable to load content.")]
    ContentLoadingError {
        #[from]
        source: ContentLoadingError,
    },

    #[error("Unable to parse template from content directory.")]
    RegisteredTemplateParseError {
        #[from]
        source: RegisteredTemplateParseError,
    },

    #[error("Unable to parse template from input.")]
    UnregisteredTemplateParseError {
        #[from]
        source: UnregisteredTemplateParseError,
    },

    #[error("Unable to render content.")]
    ContentRenderingError {
        #[from]
        source: ContentRenderingError,
    },

    #[error("Failed to write output.")]
    WriteError { source: io::Error },
}

#[derive(Error, Debug)]
pub enum GetCommandError {
    #[error("Unable to load content.")]
    ContentLoadingError {
        #[from]
        source: ContentLoadingError,
    },

    #[error("Unable to parse template from content directory.")]
    RegisteredTemplateParseError {
        #[from]
        source: RegisteredTemplateParseError,
    },

    #[error("Content not found at address '{}'.", .address)]
    ContentNotFound { address: String },

    #[error("Unable to render content.")]
    ContentRenderingError {
        #[from]
        source: ContentRenderingError,
    },

    #[error("Failed to write output.")]
    WriteError { source: io::Error },
}

/// Reads a template from `input`, renders it, and writes it to `output`.
pub fn render<I: io::Read, O: io::Write>(
    content_directory: ContentDirectory,
    source_media_type: Mime,
    target_media_type: &Mime,
    soliton_version: SolitonVersion,
    input: &mut I,
    output: &mut O,
) -> Result<(), RenderCommandError> {
    let locked_engine =
        FilesystemBasedContentEngine::from_content_directory(content_directory, soliton_version)?;
    let engine = locked_engine
        .read()
        .expect("RwLock for ContentEngine has been poisoned");

    let mut template = String::new();
    input
        .read_to_string(&mut template)
        .map_err(|source| RenderCommandError::ReadError { source })?;

    let content_item = engine.new_template(&template, source_media_type)?;
    let render_context = engine.get_render_context(target_media_type);
    let rendered_output = content_item.render(&render_context)?;
    write!(output, "{}", rendered_output)
        .map_err(|source| RenderCommandError::WriteError { source })?;

    output
        .flush()
        .map_err(|source| RenderCommandError::WriteError { source })
}

/// Renders an item from the content directory and writes it to `output`.
pub fn get<O: io::Write>(
    content_directory: ContentDirectory,
    address: &str,
    target_media_type: &Mime,
    soliton_version: SolitonVersion,
    output: &mut O,
) -> Result<(), GetCommandError> {
    let locked_engine =
        FilesystemBasedContentEngine::from_content_directory(content_directory, soliton_version)?;
    let engine = locked_engine
        .read()
        .expect("RwLock for ContentEngine has been poisoned");

    let content_item = engine
        .get(address)
        .ok_or(GetCommandError::ContentNotFound {
            address: String::from(address),
        })?;
    let render_context = engine.get_render_context(target_media_type);
    let rendered_output = content_item.render(&render_context)?;
    write!(output, "{}", rendered_output)
        .map_err(|source| GetCommandError::WriteError { source })?;

    output
        .flush()
        .map_err(|source| GetCommandError::WriteError { source })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_lib::*;
    use mime_guess::MimeGuess;
    use std::collections::{BTreeMap, HashMap};
    use std::str;

    /// Attempts to render all non-hidden files in ContentDirectory, returning
    /// them as a map of Address -> RenderedContent | ErrorMessage.
    fn render_everything(
        content_directory: ContentDirectory,
    ) -> Result<HashMap<String, String>, String> {
        let mut content = HashMap::new();
        let content_directory_root = content_directory.root();
        for content_file in content_directory {
            if !content_file.is_hidden() {
                // Create a separate ContentDirectory that can be consumed by
                // the CLI calls.
                let consumable_content_directory =
                    ContentDirectory::from_root(&content_directory_root).map_err(|error| {
                        format!("Could not create content directory: {:?}", error)
                    })?;
                let address = content_file.relative_path_without_extensions();
                let first_filename_extension = content_file.extensions().first().expect(&format!(
                    "Content file at '{}' does not have a filename extension",
                    content_file.absolute_path()
                ));

                // Target media type is just the source media type. This isn't
                // testing transcoding.
                let target_media_type = MimeGuess::from_ext(first_filename_extension)
                    .first()
                    .unwrap_or(mime::APPLICATION_OCTET_STREAM);

                let mut output = Vec::new();
                let result = get(
                    consumable_content_directory,
                    address,
                    &target_media_type,
                    SolitonVersion("0.0.0"),
                    &mut output,
                );

                let output_or_error_message = match result {
                    Ok(()) => String::from_utf8(output)
                        .map_err(|error| format!("Output was not valid UTF-8: {:?}", error))?,
                    Err(error) => {
                        let anyhow_error = anyhow::Error::from(error);
                        let causes = anyhow_error.chain().map(|error| error.to_string());
                        let message = causes.fold(String::new(), |acc, arg| acc + " " + &arg);
                        format!("Error:{}", message)
                    }
                };

                content.insert(
                    String::from(content_file.relative_path_without_extensions()),
                    output_or_error_message,
                );
            }
        }
        Ok(content)
    }

    #[test]
    fn examples_match_snapshots() {
        for content_directory in example_content_directories() {
            let content_directory_root = &content_directory.root();

            let unordered_content =
                render_everything(content_directory).expect("Fatal error in valid example");
            let contents = unordered_content
                .iter()
                // If dynamic content files mention where the repo is checked
                // out, redact it to keep tests portable.
                .map(|(key, value)| (key, value.replace(PROJECT_DIRECTORY, "$PROJECT_DIRECTORY")))
                // Files can be omitted from snapshots with a naming convention.
                .filter(|(key, _)| !key.ends_with("-SKIP-SNAPSHOT"))
                .collect::<BTreeMap<_, _>>();

            let mut insta_settings = insta::Settings::clone_current();
            insta_settings.set_input_file(content_directory_root);
            let id = content_directory_root
                .strip_prefix(
                    [PROJECT_DIRECTORY, "examples", "valid"]
                        .iter()
                        .collect::<std::path::PathBuf>(),
                )
                .or(content_directory_root.strip_prefix(
                    [PROJECT_DIRECTORY, "examples", "invalid"]
                        .iter()
                        .collect::<std::path::PathBuf>(),
                ))
                .unwrap()
                .to_string_lossy();
            insta_settings.set_snapshot_suffix(id);
            insta_settings.bind(|| insta::assert_yaml_snapshot!(contents));
        }
    }

    #[test]
    fn cli_renders_valid_templates() {
        for &(template, expected_output) in &VALID_TEMPLATES {
            let mut input = template.as_bytes();
            let mut output = Vec::new();
            let directory = arbitrary_content_directory_with_valid_content();
            let result = render(
                directory,
                mime::TEXT_HTML,
                &mime::TEXT_HTML,
                SolitonVersion("0.0.0"),
                &mut input,
                &mut output,
            );

            assert!(
                result.is_ok(),
                "Template rendering failed for `{}`: {}",
                template,
                result.unwrap_err(),
            );
            let output_as_str = str::from_utf8(output.as_slice()).expect("Output was not UTF-8");
            assert_eq!(
                output_as_str,
                expected_output,
                "Template rendering for `{}` did not produce the expected output (\"{}\"), instead got \"{}\"",
                template,
                expected_output,
                output_as_str
            );
        }
    }

    #[test]
    fn cli_fails_to_render_invalid_templates() {
        for &template in &INVALID_TEMPLATES {
            let mut input = template.as_bytes();
            let mut output = Vec::new();
            let directory = arbitrary_content_directory_with_valid_content();
            let result = render(
                directory,
                mime::TEXT_HTML,
                &mime::TEXT_HTML,
                SolitonVersion("0.0.0"),
                &mut input,
                &mut output,
            );

            assert!(
                result.is_err(),
                "Template rendering succeeded for `{}`, but it should have failed",
                template,
            );
        }
    }

    #[test]
    fn cli_can_get_content() {
        let mut output = Vec::new();
        let address = "hello";
        let expected_output = "hello world\n";

        let directory = arbitrary_content_directory_with_valid_content();
        let result = get(
            directory,
            address,
            &mime::TEXT_HTML,
            SolitonVersion("0.0.0"),
            &mut output,
        );

        assert!(
            result.is_ok(),
            "Template rendering failed for content at '{}': {}",
            address,
            result.unwrap_err(),
        );
        let output_as_str = str::from_utf8(output.as_slice()).expect("Output was not UTF-8");
        assert_eq!(
            output_as_str,
            expected_output,
            "Template rendering for content at '{}' did not produce the expected output (\"{}\"), instead got \"{}\"",
            address,
            expected_output,
            output_as_str
        );
    }

    #[test]
    fn cli_can_fail_to_get_content_which_does_not_exist() {
        let mut output = Vec::new();
        let address = "this-address-does-not-refer-to-any-content";

        let directory = arbitrary_content_directory_with_valid_content();
        let result = get(
            directory,
            address,
            &mime::TEXT_HTML,
            SolitonVersion("0.0.0"),
            &mut output,
        );

        match result {
            Ok(_) => panic!(
                "Getting content from '{}' succeeded, but it should have failed",
                address
            ),
            Err(GetCommandError::ContentNotFound {
                address: address_from_error,
            }) => assert_eq!(
                address_from_error, address,
                "Address from error did not match address used"
            ),
            Err(_) => panic!("Wrong type of error was produced, expected ContentNotFound"),
        };
    }

    #[test]
    fn cli_provides_target_media_type() {
        let mut output = Vec::new();
        let directory = ContentDirectory::from_root(&example_path("valid/media-types")).unwrap();
        let address = "echo-target-media-type";

        let media_type = mime::TEXT_HTML;

        let result = get(
            directory,
            address,
            &media_type,
            SolitonVersion("0.0.0"),
            &mut output,
        );
        assert!(
            result.is_ok(),
            "Template rendering failed for content at '{}': {}",
            address,
            result.unwrap_err(),
        );

        let output_as_str = str::from_utf8(output.as_slice()).expect("Output was not UTF-8");
        assert_eq!(
            output_as_str,
            media_type.essence_str(),
            "Template rendering for content at '{}' did not produce the expected output (\"{}\"), instead got \"{}\"",
            address,
            media_type.essence_str(),
            output_as_str
        );
    }
}
