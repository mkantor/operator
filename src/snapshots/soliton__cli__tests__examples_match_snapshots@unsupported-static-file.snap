---
source: src/cli.rs
expression: contents
input_file: examples/invalid/unsupported-static-file
---
index: "Error: Unable to load content. Content file has an unknown media type: The filename extension for the file at 'index.thisfileextensionisnotsupported' ('thisfileextensionisnotsupported') does not map to any known media type."
